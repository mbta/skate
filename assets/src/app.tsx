// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
