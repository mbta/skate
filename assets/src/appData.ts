const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
