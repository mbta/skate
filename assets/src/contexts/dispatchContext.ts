import { createContext } from "react"
import { Dispatch } from "../state"

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const DispatchContext = createContext(noopDispatch)

export default DispatchContext
