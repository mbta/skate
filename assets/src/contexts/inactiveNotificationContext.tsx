import React, { createContext, ReactElement } from "react"
import { Notification } from "../realtime.d"

export const InactiveNotificationContext = createContext<
  [Notification | null, (notification: Notification | null) => void]
>(
  // tslint:disable-next-line no-empty
  [null, () => {}]
)

export const InactiveNotificationProvider = ({
  notification,
  setInactiveNotification,
  children,
}: {
  notification: Notification | null
  setInactiveNotification: (notification: Notification | null) => void
  children: ReactElement<HTMLElement>
}) => {
  return (
    <InactiveNotificationContext.Provider
      value={[notification, setInactiveNotification]}
    >
      {children}
    </InactiveNotificationContext.Provider>
  )
}
