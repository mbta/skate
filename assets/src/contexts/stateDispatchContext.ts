import { createContext } from "react"
import { Dispatch, initialState, State } from "../state"

type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export default StateDispatchContext
