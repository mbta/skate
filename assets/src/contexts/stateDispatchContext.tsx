import React, { createContext, ReactElement } from "react"
import { Dispatch, initialState, State } from "../state"

export type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
export const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export const StateDispatchProvider = ({
  state,
  dispatch,
  children,
}: {
  state: State
  dispatch: Dispatch
  children: ReactElement<HTMLElement>
}) => {
  return (
    <StateDispatchContext.Provider value={[state, dispatch]}>
      {children}
    </StateDispatchContext.Provider>
  )
}
