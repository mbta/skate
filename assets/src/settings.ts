export enum VehicleLabelSetting {
  RunNumber = 1,
  VehicleNumber,
}

export interface Settings {
  vehicleLabel: VehicleLabelSetting
}

export const defaultSettings: Settings = {
  vehicleLabel: VehicleLabelSetting.RunNumber,
}
