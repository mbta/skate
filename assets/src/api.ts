import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
