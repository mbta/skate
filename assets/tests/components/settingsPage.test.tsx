import { mount } from "enzyme"
import React from "react"
import renderer from "react-test-renderer"
import SettingsPage from "../../src/components/settingsPage"
import { StateDispatchProvider } from "../../src/contexts/stateDispatchContext"
import { defaultSettings, VehicleLabelSetting } from "../../src/settings"
import {
  initialState,
  setLadderVehicleLabelSetting,
  setShuttleVehicleLabelSetting,
} from "../../src/state"

const mockDispatch = jest.fn()

describe("SettingsPage", () => {
  test("renders", () => {
    const tree = renderer
      .create(
        <StateDispatchProvider state={initialState} dispatch={mockDispatch}>
          <SettingsPage />
        </StateDispatchProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("displays the current ladder vehicle label setting", () => {
    const mockState = {
      ...initialState,
      settings: {
        ...defaultSettings,
        ladderVehicleLabel: VehicleLabelSetting.VehicleNumber,
      },
    }

    const wrapper = mount(
      <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
        <SettingsPage />
      </StateDispatchProvider>
    )
    const ladderVehicleLabelSelectValue = wrapper
      .find("#ladder-vehicle-label-setting")
      .prop("value")

    expect(ladderVehicleLabelSelectValue).toEqual(
      VehicleLabelSetting.VehicleNumber
    )
  })

  test("selecting a ladder vehicle label setting sets that value", () => {
    const testDispatch = jest.fn()

    const wrapper = mount(
      <StateDispatchProvider state={initialState} dispatch={testDispatch}>
        <SettingsPage />
      </StateDispatchProvider>
    )
    const testEvent = {
      currentTarget: {
        value: `${VehicleLabelSetting.RunNumber}`,
      },
    } as React.FormEvent<HTMLSelectElement>
    wrapper.find("#ladder-vehicle-label-setting").simulate("change", testEvent)

    expect(testDispatch).toHaveBeenCalledWith(
      setLadderVehicleLabelSetting(VehicleLabelSetting.RunNumber)
    )
  })

  test("displays the current map vehicle label setting", () => {
    const mockState = {
      ...initialState,
      settings: {
        ...defaultSettings,
        shuttleVehicleLabel: VehicleLabelSetting.RunNumber,
      },
    }

    const wrapper = mount(
      <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
        <SettingsPage />
      </StateDispatchProvider>
    )
    const shuttleVehicleLabelSelectValue = wrapper
      .find("#map-vehicle-label-setting")
      .prop("value")

    expect(shuttleVehicleLabelSelectValue).toEqual(
      VehicleLabelSetting.RunNumber
    )
  })

  test("selecting a map vehicle label setting sets that value", () => {
    const testDispatch = jest.fn()

    const wrapper = mount(
      <StateDispatchProvider state={initialState} dispatch={testDispatch}>
        <SettingsPage />
      </StateDispatchProvider>
    )
    const testEvent = {
      currentTarget: {
        value: `${VehicleLabelSetting.VehicleNumber}`,
      },
    } as React.FormEvent<HTMLSelectElement>
    wrapper.find("#map-vehicle-label-setting").simulate("change", testEvent)

    expect(testDispatch).toHaveBeenCalledWith(
      setShuttleVehicleLabelSetting(VehicleLabelSetting.VehicleNumber)
    )
  })
})
