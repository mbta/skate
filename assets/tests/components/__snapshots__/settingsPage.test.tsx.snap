// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsPage renders 1`] = `
<div
  className="c-page c-page--settings"
>
  <div
    className="c-page__container"
  >
    <h1
      className="c-page__title"
    >
      Settings
    </h1>
    <div
      className="c-page__section"
    >
      <h2
        className="m-settings-page__section-header"
      >
        Vehicle Settings
      </h2>
      <div
        className="m-settings-page__setting"
      >
        <div
          className="m-settings-page__setting_header"
        >
          <div
            className="m-settings-page__icon"
          >
            <span
              className="m-settings-page__icon-path"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </div>
          <div
            className="m-settings-page__setting-label"
          >
            Vehicle labels on map
          </div>
        </div>
        <div
          className="m-settings-page__options-container"
        >
          <label
            className="m-settings-page__option-label-unselected"
            data-option-id="shuttle-vehicle-label-run-number"
          >
            <input
              checked={false}
              className="m-settings-page__input"
              name="shuttle-vehicle-label"
              onChange={[Function]}
              type="radio"
              value={1}
            />
            Run number
          </label>
          <label
            className="m-settings-page__option-label-selected"
            data-option-id="shuttle-vehicle-label-vehicle-number"
          >
            <input
              checked={true}
              className="m-settings-page__input"
              name="shuttle-vehicle-label"
              onChange={[Function]}
              type="radio"
              value={2}
            />
            Vehicle number
          </label>
        </div>
      </div>
      <div
        className="m-settings-page__setting"
      >
        <div
          className="m-settings-page__setting_header"
        >
          <div
            className="m-settings-page__icon"
          >
            <span
              className="m-settings-page__icon-path"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </div>
          <div
            className="m-settings-page__setting-label"
          >
            Vehicle labels on route ladder
          </div>
        </div>
        <div
          className="m-settings-page__options-container"
        >
          <label
            className="m-settings-page__option-label-selected"
            data-option-id="ladder-vehicle-label-run-number"
          >
            <input
              checked={true}
              className="m-settings-page__input"
              name="ladder-vehicle-label"
              onChange={[Function]}
              type="radio"
              value={1}
            />
            Run number
          </label>
          <label
            className="m-settings-page__option-label-unselected"
            data-option-id="ladder-vehicle-label-vehicle-number"
          >
            <input
              checked={false}
              className="m-settings-page__input"
              name="ladder-vehicle-label"
              onChange={[Function]}
              type="radio"
              value={2}
            />
            Vehicle number
          </label>
        </div>
      </div>
      <div
        className="m-settings-page__setting"
      >
        <div
          className="m-settings-page__setting_header"
        >
          <div
            className="m-settings-page__icon"
          >
            <span
              className="m-settings-page__icon-path"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "SVG",
                }
              }
            />
          </div>
          <div
            className="m-settings-page__setting-label"
          >
            Adherence colors
          </div>
        </div>
        <div
          className="m-settings-page__options-container"
        >
          <label
            className="m-settings-page__option-label-selected"
            data-option-id="vehicle-adherence-colors-early-red"
          >
            <input
              checked={true}
              className="m-settings-page__input"
              name="vehicle-adherence-colors"
              onChange={[Function]}
              type="radio"
              value={1}
            />
            <div>
              <div
                className="m-settings-page__vehicle-adherence-setting-row"
              >
                Early bus: Red
                <div
                  className="m-settings-page__vehicle-adherence-icon"
                >
                  <span
                    className="red"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "SVG",
                      }
                    }
                  />
                </div>
              </div>
              <div
                className="m-settings-page__vehicle-adherence-setting-row"
              >
                Late bus: Blue
                <div
                  className="m-settings-page__vehicle-adherence-icon"
                >
                  <span
                    className="blue"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "SVG",
                      }
                    }
                  />
                </div>
              </div>
            </div>
          </label>
          <label
            className="m-settings-page__option-label-unselected"
            data-option-id="vehicle-adherence-colors-early-blue"
          >
            <input
              checked={false}
              className="m-settings-page__input"
              name="vehicle-adherence-colors"
              onChange={[Function]}
              type="radio"
              value={2}
            />
            <div>
              <div
                className="m-settings-page__vehicle-adherence-setting-row"
              >
                Early bus: Blue
                <div
                  className="m-settings-page__vehicle-adherence-icon"
                >
                  <span
                    className="blue"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "SVG",
                      }
                    }
                  />
                </div>
              </div>
              <div
                className="m-settings-page__vehicle-adherence-setting-row"
              >
                Late bus: Red
                <div
                  className="m-settings-page__vehicle-adherence-icon"
                >
                  <span
                    className="red"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "SVG",
                      }
                    }
                  />
                </div>
              </div>
            </div>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
